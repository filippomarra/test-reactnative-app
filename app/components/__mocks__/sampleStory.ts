import Story from '../../models';

export const sampleStory: Story = {
    kind: 'story',
    id: 151723133,
    created_at: new Date('2017-10-05T15:30:45Z'),
    updated_at: new Date('2020-09-09T08:08:04Z'),
    accepted_at: new Date('2017-10-18T14:04:39Z'),
    story_type: 'chore',
    story_priority: 'p3',
    name: 'API Design',
    description:
        'As an engineer, I will produce a RAML file to discuss with the team the entry points and the payloads for the proxy so that we can start thinking how to mock this data in the frontend',
    current_state: 'accepted',
    requested_by_id: 1945893,
    url: 'https://www.pivotaltracker.com/story/show/151723133',
    project_id: 2116794,
    owner_ids: [2958113],
    labels: [
        {
            id: 19223788,
            project_id: 2116794,
            kind: 'label',
            name: 'debito-tecnico',
            created_at: new Date('2017-10-05T15:30:45Z'),
            updated_at: new Date('2020-09-09T08:08:04Z'),
        },
    ],
    owned_by_id: 2958113,
};
