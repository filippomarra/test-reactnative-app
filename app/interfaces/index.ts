import Story from '../models';

export interface ApiProps {
    setLoading: (loading: boolean) => void;
    setStories: (stories: Story[]) => void;
}

export interface AppHeaderProps {
    stories: Story[];
    setStories: (stories: Story[]) => void;
}

export interface ModalProps {
    visible: boolean;
    hideDialog: () => void;
    story: Story;
}

export interface StoryItemProps {
    index: number;
    showDialog: (story: Story) => void;
    story: Story;
}

export interface StoryScrollViewProps {
    isLoading: boolean;
    showDialog: (story: Story) => void;
    stories: Story[];
}
