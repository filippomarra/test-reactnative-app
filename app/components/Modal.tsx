import React from 'react';
import { Linking, Text, View } from 'react-native';
import {
    Button,
    Dialog,
    Paragraph,
    Portal,
    Provider,
} from 'react-native-paper';

import dateformat from 'dateformat';

import { ModalProps } from '../interfaces';
import { styles } from '../styles';

import { DialogProperty } from './DialogProperty';

export const Modal = (props: ModalProps) => {
    const { visible, hideDialog, story } = props;

    return (
        <Provider>
            {visible && (
                <View testID="modal">
                    <Portal>
                        <Dialog
                            visible={visible}
                            onDismiss={hideDialog}
                            style={styles.dialog}>
                            <Dialog.Title>{story.name}</Dialog.Title>
                            <Dialog.Content>
                                <Paragraph>
                                    <DialogProperty text={'Description:'} />
                                    <Text>{story.description}</Text>
                                </Paragraph>
                                <Paragraph>
                                    <DialogProperty text={'URL:'} />
                                    {story.url && (
                                        <Text
                                            style={styles.url}
                                            onPress={() =>
                                                Linking.openURL(story.url!)
                                            }
                                            testID="storyUrl">
                                            {story.url}
                                        </Text>
                                    )}
                                </Paragraph>
                                <Paragraph>
                                    <DialogProperty text={'Created At:'} />
                                    {story.created_at && (
                                        <Text testID="storyDate">
                                            {dateformat(
                                                story.created_at,
                                                'dd/mm/yyyy HH:MM:ss',
                                            )}
                                        </Text>
                                    )}
                                </Paragraph>
                                <Paragraph>
                                    <DialogProperty text={'Story Type:'} />
                                    {story.story_type && (
                                        <Text testID="storyType">
                                            {story.story_type.toUpperCase()}
                                        </Text>
                                    )}
                                </Paragraph>
                                <Paragraph>
                                    <DialogProperty text={'Story Priority:'} />
                                    {story.story_priority && (
                                        <Text testID="storyPriority">
                                            {story.story_priority.toUpperCase()}
                                        </Text>
                                    )}
                                </Paragraph>
                            </Dialog.Content>
                            <Dialog.Actions>
                                <Button
                                    onPress={hideDialog}
                                    textColor="#000"
                                    testID="okButton">
                                    OK
                                </Button>
                            </Dialog.Actions>
                        </Dialog>
                    </Portal>
                </View>
            )}
        </Provider>
    );
};
