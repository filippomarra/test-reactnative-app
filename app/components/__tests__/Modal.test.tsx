import 'react-native';
import React from 'react';

import { render } from '@testing-library/react-native';

import { ModalProps } from '../../interfaces';

import { sampleStory } from '../__mocks__/sampleStory';
import { Modal } from '../Modal';

const defaultProps: ModalProps = {
    visible: false,
    hideDialog: jest.fn(),
    story: sampleStory,
};

describe('Modal component tests', () => {
    it('renders correctly with default props', () => {
        expect(render(<Modal {...defaultProps} />).toJSON()).toMatchSnapshot();
    });

    describe('When visible is set to false', () => {
        it('should not render the Modal', () => {
            const { queryByTestId } = render(<Modal {...defaultProps} />);
            expect(queryByTestId('modal')).toBeNull();
        });
    });

    describe('When visible is set to true', () => {
        it('should render the Modal', () => {
            const { queryByTestId } = render(
                <Modal {...defaultProps} visible={true} />,
            );
            expect(queryByTestId('modal')).not.toBeNull();
        });

        describe('When story url is undefined', () => {
            it('should not render the storyUrl link in the modal', () => {
                const sampleStoryCopy = { ...sampleStory };
                sampleStoryCopy.url = undefined;
                const { queryByTestId } = render(
                    <Modal
                        {...defaultProps}
                        visible={true}
                        story={sampleStoryCopy}
                    />,
                );
                expect(queryByTestId('storyUrl')).toBeNull();
            });
        });

        describe('When story url is defined', () => {
            it('should render the storyUrl link in the modal', () => {
                const { queryAllByTestId } = render(
                    <Modal {...defaultProps} visible={true} />,
                );
                expect(queryAllByTestId('storyUrl')).not.toBeNull();
            });
        });

        describe('When story type is undefined', () => {
            it('should not render the storyType link in the modal', () => {
                const sampleStoryCopy = { ...sampleStory };
                sampleStoryCopy.story_type = undefined;
                const { queryByTestId } = render(
                    <Modal
                        {...defaultProps}
                        visible={true}
                        story={sampleStoryCopy}
                    />,
                );
                expect(queryByTestId('storyType')).toBeNull();
            });
        });

        describe('When story type is defined', () => {
            it('should render the storyType link in the modal', () => {
                const { queryAllByTestId } = render(
                    <Modal {...defaultProps} visible={true} />,
                );
                expect(queryAllByTestId('storyType')).not.toBeNull();
            });
        });

        describe('When story priority is undefined', () => {
            it('should not render the storyPriority link in the modal', () => {
                const sampleStoryCopy = { ...sampleStory };
                sampleStoryCopy.story_priority = undefined;
                const { queryByTestId } = render(
                    <Modal
                        {...defaultProps}
                        visible={true}
                        story={sampleStoryCopy}
                    />,
                );
                expect(queryByTestId('storyPriority')).toBeNull();
            });
        });

        describe('When story priority is defined', () => {
            it('should render the storyPriority link in the modal', () => {
                const { queryAllByTestId } = render(
                    <Modal {...defaultProps} visible={true} />,
                );
                expect(queryAllByTestId('storyPriority')).not.toBeNull();
            });
        });
    });
});
