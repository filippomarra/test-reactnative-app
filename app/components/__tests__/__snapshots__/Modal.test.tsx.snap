// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component tests renders correctly with default props 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;
